zkvm/src/predicate.rs

```diff
@@ -136,42 +105,43 @@ impl Predicate {
         Predicate::Opaque(self.to_point())
     }
 
-    /// Creates a disjunction of a vector of predicates.
-    pub fn disjunction(preds: Vec<Predicate>) -> Result<Self, VMError> {
-        let point = {
-            let f = Predicate::commit_disjunction(preds.iter().map(|p| p.to_point()));
-            let l = preds[0]
-                .to_point()
-                .decompress()
-                .ok_or(VMError::InvalidPoint)?;
-            l + f * PedersenGens::default().B
-        };
-        Ok(Predicate::Or(PredicateDisjunction {
-            preds: preds,
-            precomputed_point: point.compress(),
-        }))
-    }
-
-    fn commit_disjunction<I>(preds: I) -> Scalar
-    where
-        I: IntoIterator,
-        I::Item: Borrow<CompressedRistretto>,
-        I::IntoIter: ExactSizeIterator,
-    {
-        let mut t = Transcript::new(b"ZkVM.predicate");
-        let iter = preds.into_iter();
-        t.commit_u64(b"n", iter.len() as u64);
-        for x in iter {
-            t.commit_point(b"X", x.borrow());
+    fn commit_taproot(key: &[u8], root: &[u8]) -> Scalar {
+        let mut t = Transcript::new(b"ZkVM.taproot");
+        t.commit_bytes(b"key", key);
+        t.commit_bytes(b"merkle", root);
+        t.challenge_scalar(b"h")
+    }
+
+    /// Verifies whether the current predicate is a commitment to a signing key `key` and Merkle root `root`.
+    /// Returns a `PointOp` instance that can be verified in a batch with other operations.
+    pub fn prove_taproot(&self, program: &Program, call_proof: &CallProof) -> PointOp {
+        let pred_leaf = PredicateLeaf::Program(program.clone());
+        let key = &call_proof.verification_key;
+        let neighbors = &call_proof.neighbors;
+        let root = MerkleTree::compute_root_from_path(b"ZkVM.taproot", &pred_leaf, neighbors);
+        let h = Self::commit_taproot(&key.0.to_bytes(), &root);
+
+        // P == X + h1(X, M)*B -> 0 == -P + X + h1(X, M)*B
+        PointOp {
+            primary: Some(h),
+            secondary: None,
+            arbitrary: vec![(-Scalar::one(), self.to_point()), (Scalar::one(), key.0)],
         }
-        t.challenge_scalar(b"f")
     }
 
-    fn commit_program(prog: &[u8], blinding: &[u8]) -> Scalar {
-        let mut t = Transcript::new(b"ZkVM.predicate");
-        t.commit_bytes(b"blinding", blinding);
-        t.commit_bytes(b"prog", &prog);
-        t.challenge_scalar(b"h")
+    /// Return predicate tree as predicate.
+    pub fn tree(
```

Comment:
let's remove this method so there's no too-thin wrappers. User can simply do `Predicate::Tree(PredicateTree::new())` and whenever it's too cumbersome, we can `impl From<PredicateTree> for Predicate` and do `PredicateTree::new().into()` instead.